pkg/image/apis/image/validation/whitelist/whitelister.go

```diff
@@ -0,0 +1,268 @@
+package whitelist
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"strings"
+
+	"github.com/golang/glog"
+
+	kerrutil "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	serverapi "github.com/openshift/origin/pkg/cmd/server/api"
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+	stringsutil "github.com/openshift/origin/pkg/util/strings"
+)
+
+// WhitelistTransport says whether the associated registry host shall be treated as secure or insecure.
+type WhitelistTransport string
+
+const (
+	WhitelistTransportAny      WhitelistTransport = "any"
+	WhitelistTransportSecure   WhitelistTransport = "secure"
+	WhitelistTransportInsecure WhitelistTransport = "insecure"
+)
+
+// RegistryWhitelister decides whether given image pull specs are allowed by system's image policy.
+type RegistryWhitelister interface {
+	// AdmitHostname returns error if the given host is not allowed by the whitelist.
+	AdmitHostname(host string, transport WhitelistTransport) error
+	// AdmitPullSpec returns error if the given pull spec is allowed neither by the whitelist nor by the
+	// collected whitelisted pull specs.
+	AdmitPullSpec(pullSpec string, transport WhitelistTransport) error
+	// AdmitDockerImageReference returns error if the given reference is allowed neither by the whitelist nor
+	// by the collected whitelisted pull specs.
+	AdmitDockerImageReference(ref *imageapi.DockerImageReference, transport WhitelistTransport) error
+	// WhitelistRegistry extends internal whitelist for additional registry domain name. Accepted values are:
+	//  <host>, <host>:<port>
+	// where each component can contain wildcards like '*' or '??' to match wide range of registries. If the
+	// port is omitted, the default will be appended based on the given transport. If the transport is "any",
+	// the given glob will match hosts with both :80 and :443 ports.
+	WhitelistRegistry(hostPortGlob string, transport WhitelistTransport) error
+	// WhitelistPullSpecs allows to whitelist particular pull specs. References must match exactly one of the
+	// given pull specs for it to be whitelisted.
+	// TODO: accept insecure flag
```

Comment:
Why is this a TODO?

Reply:
At the time I wasn't sure if this method should default ports as well for given pull spec to be consistent with other mothods. Now it's clear to me that it's not desirable since this serves to whitelist the exact string given. I'll remove the TODO line.